src/ctest.c

```diff
@@ -201,6 +214,25 @@ size_t RunTests(const TEST_FUNCTION_DATA* testListHead, const char* testSuiteNam
         (void)printf("%s%d tests ran, %d failed, %d succeeded.\n" CTEST_ANSI_COLOR_RESET, (failedTestCount > 0) ? (CTEST_ANSI_COLOR_RED) : (CTEST_ANSI_COLOR_GREEN), (int)totalTestCount, (int)failedTestCount, (int)(totalTestCount - failedTestCount));
     }
 
+#if defined _MSC_VER
+    if (std_out_handle != INVALID_HANDLE_VALUE)
+    {
+        if (SetConsoleMode_succeeded)
+        {
+            /*revert console to initial state*/
+            if (!SetConsoleMode(std_out_handle, console_mode_initial))
+            {
+                (void)printf("Error resetting console mode to initial value of %" PRIx32 ". GetLastError()=%" PRIx32 "\n", console_mode_initial, GetLastError());
+            }
+        }
+
+        if (!CloseHandle(std_out_handle))
+        {
+            (void)printf("failed to CloseHandle %p. GetLastError()=%" PRId32 "\n", std_out_handle, GetLastError());
+        }
```

Comment:
handles returned with GetStdHandle are special.
I actually did not CloseHandle on purpose based on mny reading.
They are global handles (same handle being returned at any call of GetStdHandle). Our code here is not owning that handle so it should not close it.

All MSDN examples showing usage of APIs that act on handles returned by GetStdHandle do not have CloseHandle in the examples code either (for example: [https://docs.microsoft.com/en-us/windows/console/using-the-high-level-input-and-output-functions](https://docs.microsoft.com/en-us/windows/console/using-the-high-level-input-and-output-functions)).

Here is the MSDN page that I also read which makes me think they should not have CloseHandle called on them:
[https://docs.microsoft.com/en-us/windows/console/console-handles](https://docs.microsoft.com/en-us/windows/console/console-handles)

I also did this simple console app experiment which seems to confirm my understanding:

Code:

#include <stdio.h>
#include "windows.h"

int main(void)
{
    HANDLE out_handle = GetStdHandle(STD_OUTPUT_HANDLE);
    BOOL close_result = printf("handle 1st time is %p\r\n", out_handle);
    if (!CloseHandle(out_handle))
    {
        fprintf(stderr, "Error closing handle 1st time, lasterror = %d\r\n", GetLastError());
    }
    HANDLE out_handle_2 = GetStdHandle(STD_OUTPUT_HANDLE);
    printf("handle 2nd time is %p\r\n", out_handle_2);
    if (!CloseHandle(out_handle_2))
    {
        fprintf(stderr, "Error closing handle 2nd time, lasterror = %d\r\n", GetLastError());
    }
}

Result:

handle 1st time is 00000054
Error closing handle 2nd time, lasterror = 6
Press any key to continue . . .

Reply:
Alright, alright :)

---
In reply to: [291295480](https://github.com/Azure/azure-ctest/pull/29#discussion_r291295480) [](ancestors = 291295480)